File: src/main/scala/com/pwootage/oc/js/ComponentInvoker.scala
package com.pwootage.oc.js

import com.pwootage.oc.js.jsvalue.JSValue

class ComponentInvoker {
  private var callToSync: Option[() => JSValue] = None
  private var callResult: Option[JSValue] = None

  def hasSyncCall: Boolean = this.callToSync.isDefined
  def hasResult: Boolean = this.callResult.isDefined

  def callSync(fn: () => JSValue): Unit = {
    this.callToSync = Some(fn)
  }

  def executeSync(): Unit = {
    callToSync match {
      case Some(x) => callResult = Some(x())
      case _ =>
    }
    callToSync = None
  }

  def setResult(v: JSValue): Unit = {
    callResult = Some(v)
  }

  def result(): Option[JSValue] = {
    val res = callResult
    callResult = None
    res
  }
}

File: src/main/java/com/pwootage/oc/js/ComponentInvoker.kt
package com.pwootage.oc.js

class ComponentInvoker {
  private var callToSync: (() -> JSValue)? = null
  private var callResult: JSValue? = null

  val hasSyncCall: Boolean get() = this.callToSync != null
  val hasResult: Boolean get() = this.callResult != null

  fun callSync(fn: () -> JSValue): Unit {
    this.callToSync = fn
  }

  fun executeSync(): Unit {
    if (callToSync != null) {
      callResult = callToSync?.invoke()
      callToSync = null
    }
  }

  fun setResult(v: JSValue): Unit {
    callResult = v
  }

  fun result(): JSValue? {
    val res = callResult
    callResult = null
    return res
  }
}
